### PR TITLE
add initial notes and questions for n30605

### PR DESCRIPTION
**Pull request by marcofleon**
Still working on a couple more questions, but wanted to get notes and initials q's up before the weekend.

### CODE REVIEW

File: _posts/2025-06-04-#30605.md [hunk 1]
```diff
@@ -10,17 +10,39 @@ status: upcoming
...
+    - The data structures and algorithms for cluster linearization can be found in [`cluster_linearize.h`](https://github.com/bitcoin/bitcoin/blob/master/src/cluster_linearize.h). **
+
+- The linearization code is complex and therefore relies on fuzz testing to ensure correctness. Using the wide variety of data generated by the fuzzer to build dependency graphs (`DepGraph` instances) and produce/update linearizations is probably our best bet for finding any subtle bugs or edge cases in the code.
+
+- The `cluster_linearize` fuzz test includes a few helper classes and functions that are used to compare against the actual implementation. The motivation for this PR is to construct new fuzz targets that are used to make sure these helpers work properly. By having dedicated targets that test the test, we can be even more confident that our cluster linearization code is bug-free. Woo!
+
+** There's a [PR](https://github.com/bitcoin/bitcoin/pull/32545) that aims to replace the current algorithm with a new and improved one called spanning-forest linearization. ([in-depth reading](https://delvingbitcoin.org/t/spanning-forest-cluster-linearization/1419))
```
Comment (stickies-v) [inline _posts/2025-06-04-#30605.md:24]: What should the reader take away from this line? Is it just a footnote? In which case that's worth highlighting, because it makes in-depth reading less relevant for this review club.

File: _posts/2025-06-04-#30605.md [hunk 2]
```diff
@@ -10,17 +10,39 @@ status: upcoming
...
+
+- The `cluster_linearize` fuzz test includes a few helper classes and functions that are used to compare against the actual implementation. The motivation for this PR is to construct new fuzz targets that are used to make sure these helpers work properly. By having dedicated targets that test the test, we can be even more confident that our cluster linearization code is bug-free. Woo!
+
+** There's a [PR](https://github.com/bitcoin/bitcoin/pull/32545) that aims to replace the current algorithm with a new and improved one called spanning-forest linearization. ([in-depth reading](https://delvingbitcoin.org/t/spanning-forest-cluster-linearization/1419)) 
+
+
+### Exercise (Optional)
+
+- Try fuzzing the `clusterlin_linearize` target. You can refer to [fuzzing docs](https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md) for assistance. Using `--preset=libfuzzer-nosan` will likely make it easier.
```
Comment (stickies-v) [inline _posts/2025-06-04-#30605.md:29]: Perhaps elaborating at least a bit on what "make it easier" means?

File: _posts/2025-06-04-#30605.md [hunk 3]
```diff
@@ -10,17 +10,39 @@ status: upcoming
...
+- After a few minutes, you should see a crash. Welcome to fuzzing. This is a simple example of a tiny bug in the test itself which fuzz testing catches very quickly. A lot of the time, a crash happens because something is wrong in the fuzz harness itself, as opposed to a bug in the production code. This is why it can be useful to have targets or assertions that ensure that the fuzz test itself is correct.
+
 
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+2. There are 14 targets in this fuzz test that build on each other. What is each target fuzzing? Which ones are testing the actual linearization code and which ones are testing the test?
```
Comment (stickies-v) [inline _posts/2025-06-04-#30605.md:41]: > What is each target fuzzing?

Letting all attendees describe 14 fuzz targets might be a bit much. Perhaps this might work better as some kind of puzzle, like "Which target is doing X"?

File: _posts/2025-06-04-#30605.md [hunk 4]
```diff
@@ -10,17 +10,39 @@ status: upcoming
...
+    - The data structures and algorithms for cluster linearization can be found in [`cluster_linearize.h`](https://github.com/bitcoin/bitcoin/blob/master/src/cluster_linearize.h). **
+
+- The linearization code is complex and therefore relies on fuzz testing to ensure correctness. Using the wide variety of data generated by the fuzzer to build dependency graphs (`DepGraph` instances) and produce/update linearizations is probably our best bet for finding any subtle bugs or edge cases in the code.
+
+- The `cluster_linearize` fuzz test includes a few helper classes and functions that are used to compare against the actual implementation. The motivation for this PR is to construct new fuzz targets that are used to make sure these helpers work properly. By having dedicated targets that test the test, we can be even more confident that our cluster linearization code is bug-free. Woo!
+
+** There's a [PR](https://github.com/bitcoin/bitcoin/pull/32545) that aims to replace the current algorithm with a new and improved one called spanning-forest linearization. ([in-depth reading](https://delvingbitcoin.org/t/spanning-forest-cluster-linearization/1419)) 
+
+
+### Exercise (Optional)
```
Comment (stickies-v) [inline _posts/2025-06-04-#30605.md:27]: Adding a question on this exercise might be nice, something like "Were you able to run the `clusterlin_linearize` fuzz test? How long did it take to produce a crash after making the `s/chunking/simple_chunking/` change?"

**Review by stickies-v — COMMENTED** [review]
Thanks @marcofleon, notes look like they're in good shape and imo comprehensive enough, so probably mostly good to add a few more questions?

Made 2 small edit commits but don't seem to have access to your branch (maybe because it's your master branch, usually best to use separate branch for this stuff), so if you like them you can pick them yourself: https://github.com/stickies-v/website/commits/pr/834/

**Comment by marcofleon** [comment]
Thank you for the feedback @sipa and @stickies-v. I believe I addressed it all. If you have any more comments or recs, please let me know.

**Comment by marcofleon** [comment]
Also @stickies-v would you prefer that I squash the commits or is it fine as is?

**Review by stickies-v — APPROVED** [review]
Pushed one small edit on SFL, but otherwise LGTM! Will merge in 2 hours unless more comments come in.

File: _posts/2025-06-04-#30605.md [hunk 5]
```diff
@@ -7,20 +7,52 @@ authors: [sipa]
...
+
+6. In `clusterlin_simple_finder`, when finding a topologically valid subset to remove from the graph, why is it important to set `non_empty` to true? What could happen if we allowed empty sets?
+
+7. In `clusterlin_simple_linearize`, why does the code only verify against all possible permutations when `depgraph.TxCount() <= 8`? What would happen if we tried to do this for larger transaction counts?
+
+8. In [commit 1e4f345](https://github.com/bitcoin-core-review-club/bitcoin/commit/1e4f3452a980744ece8a57bf46c35f44e2aa83a0), when a non-topological permutation is found, the code now fast forwards by reversing part of the permutation. Why does this optimization work, and how many permutations can it skip in the best case?
+
+9. In `SimpleCandidateFinder::FindCandidateSet()`, the algorithm keeps track of two sets for each work unit: `inc` (transactions definitely included) and `und` (transactions that can still be added). Why does it need to track both sets instead of just tracking undecided transactions?
+
+10. What is, no doubt, the coolest way to test code?
```
Comment (glozow) [inline _posts/2025-06-04-#30605.md:55]: lol :joy:
